### PR TITLE
[BDHK-265] Simulate TX before submitting onchain with Tenderly

### PR DESCRIPTION
**Pull request by adelowo**
Required env values:

```
ERC4337_BUNDLER_TENDERLY_ENABLE_SIMULATION=true
ERC4337_BUNDLER_SIMULATION_DURATION=10
```

This PR implements a safeguard to verify the functionality of the solved calldata we are about to execute onchain.

## Summary by CodeRabbit

- **New Features**
	- Introduced transaction simulation capabilities using the Tenderly API, allowing for improved validation before execution.
	- Enhanced configuration options for simulation, including enabling/disabling features and setting timeouts through environment variables.

- **Bug Fixes**
	- Improved error handling during gas estimation for individual user operations, enhancing robustness and responsiveness.

### CODE REVIEW

File: srv/tenderly.go [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+						Solved: currentOpHash,
+					}, false
+				})
+
+				return err
+			}
+
+			outgoingAssetChanges := simulatedResponse.Result.AssetChanges[0]
+
+			if common.HexToAddress(outgoingAssetChanges.From).Hex() != userop.Sender.Hex() {
```
Comment (blewater) [inline srv/tenderly.go]: Good!

File: srv/tenderly.go [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+func SimulateTxWithTenderly(cfg *conf.Values,
+	chainCliet *ethclient.Client,
+	logger logr.Logger, txHashes *xsync.MapOf[string, OpHashes],
+	entrypointAddr common.Address, chainID *big.Int) modules.BatchHandlerFunc {
+
+	client := &http.Client{
+		Timeout: cfg.SimulationTimeout,
+	}
+
+	return func(ctx *modules.BatchHandlerCtx) error {
```
Comment (blewater) [inline srv/tenderly.go:33]: Long, you can break this.

File: srv/tenderly.go [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+					}, false
+				})
+
+				return err
+			}
+
+			incomingAssetChanges := simulatedResponse.Result.AssetChanges[len(simulatedResponse.Result.AssetChanges)-1]
+
+			if common.HexToAddress(incomingAssetChanges.To).Hex() != userop.Sender.Hex() {
+				err = errors.New("outward transaction does not belong to the userop sender")
```
Comment (blewater) [inline srv/tenderly.go]: outward `userOp` the bundler account sends the transaction.

File: srv/tenderly.go [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+		} `json:"assetChanges"`
+		BalanceChanges []struct {
+			Address     string `json:"address"`
+			DollarValue string `json:"dollarValue"`
+			Transfers   []int  `json:"transfers"`
+		} `json:"balanceChanges"`
+	} `json:"result"`
+}
+
+func SimulateTxWithTenderly(cfg *conf.Values,
```
Comment (blewater) [inline srv/tenderly.go:24]: Can an integration test with a mock userOp?

**Review by blewater — CHANGES_REQUESTED** [review]
Ask Ofir whether to expose this as an RPC endpoint for the UI.

File: conf/values.go [hunk 5]
```diff
@@ -153,6 +159,10 @@ func GetValues() *Values {
 	_ = viper.BindEnv("solver_url")
 	_ = viper.BindEnv("erc4337_bundler_status_timeout")
 	_ = viper.BindEnv("erc4337_bundler_address_whitelist")
+	_ = viper.BindEnv("erc4337_bundler_tenderly_enable_simulation")
+	_ = viper.BindEnv("erc4337_bundler_tenderly_url")
+	_ = viper.BindEnv("erc4337_bundler_tenderly_access_key")
```
Comment (hpsing) [inline conf/values.go:164]: we don't need `access_key` if we using `tenderly_simulateTransaction` RPC call instead API

**Review by blewater — APPROVED** [review]
Great!

**Pull request merged**